### PR TITLE
CLI - `build` command has a param for wasm-opt flags

### PR DESCRIPTION
**Pull request by bfops**
# Description of Changes

Add a `spacetime build --wasm-opt=false` option, so users can skip wasm-opt even if they have it installed.

Note that this is mutually exclusive with `--debug`. I could be persuaded to merge `--wasm-opt` and `--debug` into a `--profile=debug|release|release-no-wasm-opt` flag if that's preferred.

It also doesn't error if you provide `--wasm-opt=true` and then it can't find wasm-opt. Should that be an error instead of a "warning"? (And the default would become "wasm-opt-if-available" or something).

Addresses https://github.com/clockworklabs/SpacetimeDB/issues/846.

# API and ABI breaking changes

No. This is a new param.

# Expected complexity level and risk

2

# Testing

*Describe any testing you've done, and any testing you'd like your reviewers to do,
so that you're confident that all the changes work as expected!*

- [x] *Write a test you've completed here.*
- [ ] *Write a test you want a reviewer to do here, so they can check it off when they're satisfied.*

### CODE REVIEW

File: crates/testing/tests/standalone_integration_test.rs [hunk 1]
```diff
@@ -43,39 +43,34 @@ async fn read_logs(module: &ModuleHandle) -> Vec<String> {
 fn test_calling_a_reducer_in_module(module_name: &'static str) {
     init();
 
-    CompiledModule::compile(module_name, CompilationMode::Debug).with_module_async(
```
Comment (bfops) [inline crates/testing/tests/standalone_integration_test.rs:46]: clippy really wanted to reformat this. It helps to review with whitespace hidden, but it's still not great.

File: crates/cli/src/subcommands/build.rs [hunk 2]
```diff
@@ -30,12 +31,21 @@ pub fn cli() -> clap::Command {
...
                 .help("Builds the module using debug instead of release (intended to speed up local iteration, not recommended for CI)"),
         )
+        .arg(
+            Arg::new("wasm_opt")
+                .long("wasm-opt")
+                .value_parser(clap::builder::BoolishValueParser::new())
+                .default_value("true")
+                .conflicts_with("debug")
+                .help("Whether to optimize the compiled module with wasm-opt (default: true)"),
+        )
```
Comment (bfops) [inline crates/cli/src/subcommands/build.rs:41]: I could also be persuaded to merge `--wasm-opt` and `--debug` into a `--profile=debug|release|release-no-wasm-opt` flag if that's preferred.

**Comment by RReverser** [comment]
> Add a `spacetime build --wasm-opt=false` option, so users can skip wasm-opt even if they have it installed.

Hm what's the usecase? If someone wants quick iteration, I'd say they should use debug builds (which already don't run wasm-opt), and if someone wants to publish a fast module, they should always run wasm-opt as it yield noticeable boost.

**Comment by bfops** [comment]
> > Add a `spacetime build --wasm-opt=false` option, so users can skip wasm-opt even if they have it installed.
> 
> Hm what's the usecase? If someone wants quick iteration, I'd say they should use debug builds (which already don't run wasm-opt), and if someone wants to publish a fast module, they should always run wasm-opt as it yield noticeable boost.

See https://github.com/clockworklabs/SpacetimeDB/issues/846

**Comment by RReverser** [comment]
> at least under the default wasm-opt config, strips their debug symbols, making profiler output (e.g. perf) incomprehensible.

IIRC the original motivation is no longer relevant as we now correctly pass a flag to preserve debug info in wasm-opt (wasm-opt has always supported preserving debug symbols, we just didn't pass that flag originally).

**Comment by bfops** [comment]
> > at least under the default wasm-opt config, strips their debug symbols, making profiler output (e.g. perf) incomprehensible.
> 
> IIRC the original motivation is no longer relevant as we now correctly pass a flag to preserve debug info in wasm-opt (wasm-opt has always supported preserving debug symbols, we just didn't pass that flag originally).

I have no context - I'd love for this to be true! All I know is I was assigned this ticket last week :shrug: I'll ping @cloutiertyler and/or @gefjon.

**Comment by RReverser** [comment]
Just for the context, the fix I'm referring to was here, about a month after the original issue. https://github.com/clockworklabs/SpacetimeDB/pull/1013

**Comment by gefjon** [comment]
So long as we're preserving debug symbols, I consider the original issue solved and see no need to remove wasm-opt. I also don't see any downside to providing the flag, but certainly it's not a priority. Long-term I'd like to have a mechanism for passing arbitrary wasm-opt flags through `spacetime build`/`spacetime publish`, but again, not a priority.

**Comment by RReverser** [comment]
> Long-term I'd like to have a mechanism for passing arbitrary wasm-opt flags

Yeah I think long-term we should have something like `spacetime.toml` where we can specify all sorts of configs, including this one. Command-line args can only hold so much.

**Pull request closed without merging**